posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,285 @@
+
+  // My protocol + the responses I need to understand...
+  private type CommandsAndResponses = Command | PingPong.Response
+
+  def apply(pingPong: ActorRef[PingPong.Ping]): Behavior[Command] = {
+    val internalBehavior: Behavior[CommandsAndResponses] =
+    Behaviors.setup { context =>
+      Behaviors.receiveMessage {
+        case StopPingPong =>
+          context.log.info(s"End of the ping-pong game - calling it a day!")
+          context.system.terminate()
+          Behaviors.stopped
+        case SendPing =>
+          pingPong ! PingPong.Ping(replyTo = context.self)
+          Behaviors.same
+        case response : PingPong.Response =>
+          context.log.info(s"Hey: I just received a $response !!!")
+          Behaviors.same
+      }
+    }
+    internalBehavior.narrow
+  }
+}
+----
+
+Admittedly, some magic has happened here and the question to ask is, _What is this `narrow` method doing_? Let's look at that in the next section.
+
+== Clearing up some magic
+
+The `Behavior` abstract class in Akka 2.6 defines the `narrow` method and here is a part of the relevant source code:
+
+[source,scala,linenums]
+----
+abstract class Behavior[T](private[akka] val _tag: Int) { behavior =>
+
+  /**
+   * Narrow the type of this Behavior, which is always a safe operation. This
+   * method is necessary to implement the contravariant nature of Behavior
+   * (which cannot be expressed directly due to type inference problems).
+   */
+  final def narrow[U <: T]: Behavior[U] = this.asInstanceOf[Behavior[U]]
+  
+  ...
+}
+  
+abstract class ExtensibleBehavior[T] extends Behavior[T](BehaviorTags.ExtensibleBehavior) {
+
+  ...
+  
+  def receive(ctx: TypedActorContext[T], msg: T): Behavior[T]
+  
+  ...
+----
+
+There's quite a bit going on here.
+
+First, we note that the `Behavior` class is generic: it has a type parameter `T`, which, because of no specific variance annotation on `T`, implies that `Behavior` is _invariant_ in its type parameter `T`. Also note the comment on the `narrow` method, stating: 
+
+> _This method is necessary to implement the contravariant nature of Behavior (which cannot be expressed directly due to type inference problems)_.
+
+Second, we see that the class `ExtensibleBehavior`, which is a subclass of `Behavior`, has a `receive` method which takes a parameter `msg` of type `T`. Because functions (or _methods_), are contravariant in their argument types, the only possible variance case for the type parameter is invariant (`T`) or contravariant (`-T`). Because of type inference problems in Scala 2, the former was chosen.
+
+NOTE: Variance manifests itself in specific contexts and is a topic in itself with _contravariance_ being the least intuitive. We'll see however that, in the case of `Behavior`, it is actually quite easy to understand. I'll get back to that later. For a comprehensive explanation of variance in Scala read https://docs.scala-lang.org/tour/variances.html[this article].
```

Comment:
I think the first sentence could be cleared up a bit. I'm not 100% sure what you mean wen you say it's a "topic in itself". Do you mean it's a large topic? Do you mean it's a difficult topic with contravariance being the least intuitive?

Reply:
Reading it over, you could even just add an adjective in front of topic to clarify.

Reply:
Added an adjective